### DISCUSSION TITLE
Fast method of drawing round points

### DISCUSSION THREAD

**Discussion by bforl** [mono/SkiaSharp#2819] [Q&A]
I'm working on a custom WPF control that displays a scatter point chart, showing upwards of 200k points. This chart is interactive, meaning the user can pan, scroll, zoom etc and the rendering will update.

However, the performance when trying to render such a large number of points is pretty poor, and the interactive element is ... less interactive than I would like (as its so laggy)

I am currently using SKCanvas.DrawPoints, and the documentation specifically mentions that points are rendered "one at a time". I was wondering if there is some other method I could be using? perhaps something that uses some kind of geometry instancing? (i.e one draw call, multiple data)

Some information about my use case

- The list of points is static, but when panning/scrolling/zooming, their relative position on screen changes, so whilst the data is static, the position on screen is dynamic. 
- the size of each point is the same
- colours are all the same
- needs to support opacity

Some things I have noticed :
- rendering as a square instead of a circle is MUCH faster (but would prefer round)
- IsAntialias makes little to no difference

One option would be to dynamically to switch from round to square depending on the number of points.

Here is some example code for what I am doing. 

**NOTE : The memory allocation of creating the points each time is not the bottleneck, it is the render call that takes all the time. But I have written it like this as it keeps the example code simple.**

```
private void OnPaintSurface(object sender, SKPaintSurfaceEventArgs e)
{
    var canvas = e.Surface.Canvas;
    canvas.Clear(SKColors.White);

    using var paint = new SKPaint
    {
        Color = SKColors.Black.WithAlpha(50),
        IsAntialias = false,
        Style = SKPaintStyle.Fill,
        StrokeCap = SKStrokeCap.Round,
        StrokeWidth = 6,
    };

    var points = new List<SKPoint>();
    var r = new Random(0);

    for (int i = 0; i < 200_000; ++i)
    {
        points.Add(new SKPoint((float)(r.NextSingle() * ActualWidth), (float)(r.NextSingle() * ActualHeight)));
    }

    canvas.DrawPoints(SKPointMode.Points, points.ToArray(), paint);
}
```

**Comment by bforl** [comment]
Any suggestions? :)

**Reply by malstraem** [reply]
1. Draw 200k point with round and stroke (i.e. using Skia internal fragment shader) is not something you should do with libraries like Skia. Perhaps modern GPU have no problems with this case, but OpenGL (and Skia entire architecture) makes it expensive I think.
2. How its looks? 200k real data points are unreadable anyway, most of them overlap each other on base scale. So you need to filter or/and cluster data using some tree structure and cull it when zoom or movement occurs.

**Reply by bforl** [reply]
Yep, I agree on both points, but I still wanted to ask anyway as I thought there might be some low level options available.

**Reply by taublast** [reply]
If i had such task i would consider one of the following:
1. check if creating some SKPath with all those points and then drawing it once is a viable solution (no idea)
2. use skiasharp 3 preview and A. use a hardware accelerated view B. draw all those points using a shader.
In my opinion the second option will solve the task 100%.

**Reply by bforl** [reply]
Thanks, I'll try the 2nd option and see how I go. Do you have any example code to get me started?

In regards to the first option, that turned out to be a lot slower.

**Reply by taublast** [reply]
You can check out how a shader is created/consumed in skiasharp3 here, you can just grab the code from those wrappers and use in your app directly. If you decide to compile both projects for skia3 (demo and engine) both must have for both inside .csproj `<UseSkiaSharp3>true</UseSkiaSharp3>` as by default its `false`.
https://github.com/taublast/DrawnUi.Maui/blob/f81d4ece68dc23c8188a55af0c100edb96232007/src/samples/Sandbox/Views/Controls/ContentFolder.cs#L180

For more details about skia SKSL syntax shaders: 
https://skia.org/docs/user/sksl/
https://shaders.skia.org/

At the same time I'm sure you could find already existing SKGL shader for drawing such points and convert it to SKSL.

**Reply by bforl** [reply]
Just to clarify, are you talking about generating the point data on the fly? (like the examples you show where they're using screen space effects)

Or do you mean I could still pass in the 200k points for the shader to render? Which is what I need as the points are statistical data that I need to plot. I can't procedurally create the data as the shaders you link to are doing.

**Reply by taublast** [reply]
A. If you compute points coords yourself, it might be better better to initially send to shader some lookup table/other data used in computation and compute everything on GPU then render the shader result.
B. If you have some realtime data, like other said, there are points that are [not visible]/[can be combined into one visible for that pixel] so you will not want to pass 200k for everyframe, usually graphics engines render visible area only.
C. In case of never changing 200k points coords you might send that lookup table just once then render visible/aggregated area using shader code.

**Reply by bforl** [reply]
A : As mentioned in the original question, its a scatter plot of statistical data, I am not computing it on the fly.
B : I really do want to render all the points. It's a plot with transparency.
![image](https://github.com/mono/SkiaSharp/assets/7289277/ea2911b0-c2de-4794-8b79-d83d4c716668)


Do you have an example that does somthing like C where its dealing with an array of data? The shader examples I have found are only for dealing with small amounts of data and only send a few floats as params.

**Reply by mattleibow** [reply]
I still think the issue is that you are drawing 200k points. If you have a view that is 400x400 and draw a single point in each pixel, you still can only draw 160k points before you start overlapping.

I also think that maybe "don't draw so many points" is ambiguous. I am meaning don't do 200k draw calls as that is a waste. Most likely you can give the illusion of 200k by merging nearby points. Especially since the size of each dot is 6px. This means you are most likely drawing multiple times over. 

Do you have a sample of real data? I can have a quick play around and also do benchmarking.

How long does a single frame take to render with the CPU/raster and how long with the GPU/GL? If you use the profiling tools in VS, where is the main time?

**Reply by mattleibow** [reply]
Here is an example of a 800x800 canvas with 200k random points: 
![image](https://github.com/user-attachments/assets/572fffe9-f429-446b-a05e-c4b9dae28373)

**Reply by mattleibow** [reply]
I wrote the simplest filter ever that just checks to see if any points are close together and I went from 500ms to 25ms with 200k points:

```cs
var filtered = new Dictionary<SKPoint, SKPoint>(points.Count);
foreach (var point in points)
{
    var key = new SKPoint((int)(point.X / 6), (int)(point.Y / 6));
    filtered[key] = point;
}

return filtered.Values.ToArray();
```

This was still slow because all the points are evenly distributed across the board, but with a real dataset, this may even be far less as most of the time they are located in chunks. 

This does filter evenly and thus density is lost, but if we are to reduce the data set, we could instead have some sort of density-based filter where it keeps more points in dense areas and loses more in less areas - exaggerating the graph.

In many cases, the user can't really make out data if all 200k points are presented as it will just look like a blob. Using exaggeration will help with data presentation.

**Reply by mattleibow** [reply]
Testing with the `SKGLElement` view, I can draw things 10x faster. All 200k points at 50ms instead of 500ms and the filtered one at 7ms instead of 25ms.

So definitely use that.

**Reply by bforl** [reply]
As already mentioned, its a scatter plot with a large number of points. Opacity is used to see where build up of points occur for further filtering. So overlapping points is by design. 

(This is the type of thing, although this is showing fewer points)
![image](https://github.com/user-attachments/assets/3f72f392-13b5-4cc4-8af9-f5963f41eb60)

Yes - its a large number of points, but its a large data set :) (users can zoom in and out to explore the data)

I have since figured out a more efficient of way rendering using SIMD. I take the SKSurface and get a pointer to it's backbuffer, and then draw my own points using AVX instructions. It's about 4x quicker (10x quicker if I sort the points by their Y position).

So am I now rendering 256k points to a RGBA 3500x1400 back buffer in 5ms (vs 95ms with DrawAtlas/DrawPoints)

The main thing I have learnt though is that the bottleneck is not so much the rendering (although that definitely plays a part), but the memory usage. With such a large image buffer, there are a lot of cache misses.

All the above is with SKElement.

in regard to SKGLElement, I find it can certainly be quicker when you only have one, but if you have multiple (e.g a dashboard), the performance is goes right down, and CPU rendering becomes quicker.

**Comment by mattleibow** [comment]
I think I may be seeing some potential issues:

 - the `.ToArray()` is far more expensive than it looks because each call is allocating a new array instance, consuming memory and then causing the GC to collect.
 - there are GPU views that may result in a faster render time - in SkiaSharp 3.x there is a new SKGLElement that we just added
 - there may be unnecessary points

For the first option, I know there is no alternative API and we need to add one that uses a span. However, even with the nice fast span, the backing/source collection also needs to be accessible via an array - this may help: https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan

If you try the new GPU-backed view in 3.x, is it faster?

Besides all this, trying to draw 200k points is also sometimes less useful if the points are off the screen. It may be worth breaking the points up into regions. For example, if the total chart is holding points in a visible region of 500x300px, then it may be useful to break the points into boxes of 100x100px and just drawing those. This does not solve the case where all the points are visible, but it may help with panning.

You may also want to look at merging points at very zoomed out levels. Instead of drawing 7 points in a single pixel or even a 3x3px area, just draw one. This will involve processing the point data, but drawing less is always faster.

I know some of this does not help you now, but the GPU accelerated view may help. If the point collection is fixed, then it may be worth creating the array one time to pass to the points method, then it does not matter about array vs span vs anything, there is a single, long-lived array instance.

**EDIT:** I also just realized the `.ToArray()` may just be for example purposes - but also may not. So still valid.

**Comment by bforl** [comment]
Thanks

I did specifically mention the array allocations in the original question (I tested with a static array and it made no noticeable difference).

And yep, I recently tried SKGLElement, but sadly, made no noticeable difference.

I agree with all your other points about approaching the problem from another angle by simplifying the point collection, and that is something I will probably do, but it was more a question to see if there was anything lower level I could use. 200k might sound like a lot, .. but by todays hardware really isn't that much work (even if the points are polygons).

Im guessing that each point is tessellated on the CPU? as opposed to it being tessellated on the GPU?

And of course, I appreciate that I am probably expecting too much from this library, (trying to use it for something it wasn't intended for) which of course is fair enough :)

**Reply by mattleibow** [reply]
This may be a question for the skia-discuss group as there may be a totally different way to do something like this: https://groups.google.com/g/skia-discuss

**Comment by artemiusgreat** [comment]
If you don't need to support Web UI, you could use OpenGL library instead, e.g. this one 
https://github.com/artemiusgreat/GLGraph
